url8b434243b7_ondrasej_GPlan_tgz-pJ8-PlanningProblemTestJ8/src/cz/matfyz/sykora/planning/predicate/PredicateParser.java

```diff
@@ -0,0 +1,754 @@
+	 *	@throws TokenException on invalid input data format.
+	 *	@see #parse()
+	 *	@see #parseAction(Hashtable)
+	 *	@see #parseAction(String, Hashtable)
+	 */
+	public Action parseAction() throws IOException, TokenException {
+		return parseAction(new Hashtable<String, ValueBinding>());
+	}
+	
+	/**
+	 * Reads action specification from the stream. Uses a given set of variable
+	 * bindings.
+	 * 
+	 *	@param _variables list of variable bindings
+	 *	@return a new action based on data from the input reader. 
+	 *	@throws IOException on input/output error.
+	 *	@throws TokenException on invalid input data format.
+	 *	@see #parse()
+	 *	@see #parseAction()
+	 *	@see #parseAction(String, Hashtable)
+	 */
+	public Action parseAction(Hashtable<String, ValueBinding> _variables) throws IOException, TokenException {
+		// parse action name
+		Token current = nextToken();
+		if(current == null)
+			return null;
+		if(current.getTokenType() != Token.Type.IDENTIFIER)
+			throw new UnexpectedTokenException(current, "Identifier (action name) expected");
+		
+		String action_name= current.getTokenText();
+		
+		// parse rest of the action data
+		return parseAction(action_name, _variables);
+	}
+	
+	/**
+	 * Reads action specification from the stream. Uses a given set of variable
+	 * bindinges.
+	 * 
+	 * This version does not read the identifier specifying name of the action.
+	 * Thus it can be used from within the method parse which needs to process
+	 * first two tokens of the object itself.
+	 * 
+	 *	@param _action_name name of the action.
+	 *	@param _variables list of variable bindings that should be used to
+	 * 		translate variable names to it's inner representation.
+	 *	@return a new instance of action based on the data from the input
+	 * 		stream.
+	 *	@throws IOException on input/output error.
+	 *	@throws TokenException on invalid input data format.
+	 *	@see #parse
+	 *	@see #parseAction(Hashtable)
+	 */
+	protected Action parseAction(String _action_name,
+			Hashtable<String, ValueBinding> _variables) throws IOException, TokenException {
+		
+		// verify there is a quad dot after the action name 
+		Token current = nextToken();
+		if(current == null)
+			throw new UnexpectedEndOfInputException();
+		if(current.getTokenType() != Token.Type.QUADDOT)
+			throw new UnexpectedTokenException(current, "Quad dot expected");
+	
+		// parse preconditions
+		PredicateSet preconditions = new PredicateSet();
+		
+		current = nextToken();
+		if(current == null)
+			throw new UnexpectedEndOfInputException();
+		while(current.getTokenType() == Token.Type.IDENTIFIER
+				|| current.getTokenType() == Token.Type.COMMA) {
+			if(current.getTokenType() == Token.Type.COMMA) {
+				current = nextToken();
+				if(current == null)
+					throw new UnexpectedEndOfInputException();
+				if(current.getTokenType() != Token.Type.IDENTIFIER)
+					throw new UnexpectedTokenException(current, "Predicate starting with an identifier expected");
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 501:507, 539:545.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.